### ISSUE TITLE
[selectors-4] Whitespace aware attribute-substring matching

### ISSUE THREAD

**Issue by jakob-e** [w3c/csswg-drafts#8124]
The current [attribute substring](https://www.w3.org/TR/selectors-4/#attribute-substrings) allows matching start and end of value 

```
[attr^="value"]

// Will match 
1 <div attr="value">
2 <div attr="value list">
3 <div attr="values">
4 <div attr="value-list">
```

To add precision the substring start/end (^$) could be combined with the whitespace-separated flag (~) to ensure an exact match of the substring 
```
[attr^~="value"]

// Would only match
1 <div attr="value">
2 <div attr="value list">
```

**Syntax suggestion**
```
[attr^~="value"] { .. }
[attr$~="value"] { .. }
```

**_Spec:_** 
https://www.w3.org/TR/selectors-4/#attribute-representation
https://www.w3.org/TR/selectors-4/#attribute-substrings

**Comment by Loirooriol** [comment]
What's your usecase? I think that space-separated values are typically unordered. And if they are ordered, just allowing to check for the 1st and last position may not be doing much.

**Comment by jakob-e** [comment]
I've changed the title to perhaps better illustrate my intend

A use case is targeting components and variations without collisions

```
[data-comp^~="card"] { ... }                     /* card component */ 
[data-comp^~="card"][data-comp~="news"] { ... }  /* news card variation */ 
[data-comp^~="card"][data-comp~="fancy"] { ... } /* fancy card variation */ 

// would match (in order)
<article data-comp="card">
<article data-comp="card news">
<article data-comp="card fancy">
```

The above syntax will allow creating 
```
[data-comp^~="card-header"] { ... } /* card header component */
<article data-comp="card-header"> 
```
...**without** the latter also being matched by the _card component selector_ (as it will - using the existing `^=` syntax) 

The reason I picked the whitespace `~` for my example syntax was to convey the same idea of bounding (exact match)

**Comment by tabatkins** [comment]
Yeah, it's not uncommon to have this ordered (or perhaps partially ordered, with the first token treated specially and the rest unordered) sort of token representation in design systems. The `^=` misleads you a bit in this regard, leading to the exact mistake @jakob-e points out, where a `[attr^="card"]` ends up misfiring on "card-foo" subtypes.

**Comment by Crissov** [comment]
I strongly believe it is an anti-pattern to special-case the first entry in a space (or comma) separated list of entries in an attribute value. Even though it exists, Selectors should therefore not add special syntax for matching it. Authors should use two attributes instead. 

For hierarchical attribute values, we do have a convention already: the top-most level is prefixed to the lower level with a hyphen between them. The top level can be matched by `|=`, i.e. `[foo|=bar]` is equivalent to `:is([foo=bar], [foo^=bar-])`.

**Comment by jakob-e** [comment]
@Crissov 
It's been a while -  my suggestion is about having an **_exact match for the first token_** (or last) 
to ease e.g. component styling without collisions. 


```html 

Using |= won't work for exact match on first token 
[data-comp|=foo] { background: green; }

<div data-comp="foo"></div>         ✅ will be green (wanted)
<div data-comp="foo bar"></div>     ❌ will not be green (unwanted)
<div data-comp="foo-bar baz"></div> ❌ will be green (unwanted foo != foo-bar)

Using ^= won't work either 
[data-comp^=foo] { background: green; }

<div data-comp="foo"></div>         ✅ will be green (wanted)
<div data-comp="foo bar"></div>     ✅ will be be green (wanted)
<div data-comp="foo-bar baz"></div> ❌ will be green (unwanted foo != foo-bar)


Using ~= will also fail
[data-comp~=foo] { background: green; }

<div data-comp="foo"></div>         ✅ will be green (wanted)
<div data-comp="bar foo"></div>     ❌ will be be green (unwanted foo is not first token)
<div data-comp="foo-bar baz"></div> ✅ will not be green (wanted foo != foo-bar)

```

The current solution is to add a exact match combined a starts with + space (for variations) 
`:is([data-comp="foo"],[data-comp^="foo "]) { ... }` 


(reaction is not ment in a harsh way 🤗).

**Comment by Crissov** [comment]
`[data-comp~="foo"][data-comp^="foo "]` should also match the right subset. 

I therefore agree in so far that either `[attr^~=val]` or `[attr~^=val]` would be an intuitive syntax for this attribute value selector.

**Comment by jakob-e** [comment]
That also works with a slightly higher specificity. In regards to syntax I don’t have strong opinions about A or B  ^~= probably read how I would explain it ‘starts with and bounded by whitespace’ but it could also be a third option I haven’t thought about
